docs/src/main/asciidoc/rest-json.adoc

```diff
@@ -245,6 +245,151 @@ public class CustomObjectMapper {
 }
 ----
 
+There is one unqualified `ObjectMapper` instance provided by the quarkus-jackson extension.  Your code may produce two additional
+`ObjectMapper` instances. The client `ObjectMapper` instance is qualified by the `@RestClientObjectMapper` annotation. The server
+`ObjectMapper` instance is qualified by the `@RestServerObjectMapper` annotation.
+
+If the client `ObjectMapper` instance is available, it is used by the quarkus-jaxrs-client-reactive extension for serialization
+and deserialization.  If the client instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
+instance.
+
+If the server `ObjectMapper` is available, it is used by the quarkus-resteasy-reactive extension for serialization and
+deserialization.  If the server instance is not provided, then quarkus-resteasy-reactive uses the unqualified `ObjectMapper`
+instance.
+
+You can produce, inject, and customize each of these instances independently.  The following snippet demonstrates how to inject
+each instance using the `@RestClientObjectMapper` and `@RestServerObjectMapper` annotations.
+
+[source,java]
+----
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.quarkus.resteasy.reactive.jackson.common.RestClientObjectMapper;
+import io.quarkus.resteasy.reactive.jackson.common.RestServerObjectMapper;
+
+import javax.inject.Inject;
+
+public class UseFallback {
+
+    private final ObjectMapper serializer;
+
+    @Inject
+    public UseFallback(ObjectMapper mapper,
+            @RestClientObjectMapper Instance<ObjectMapper> clientInstance) {
+        serializer = clientInstance.isUnsatisfied() ? mapper : clientInstance.get();
+    }
+
+}
+----
+
+If you want to customize the unqualified, client, and server instances differently, you can use the `@RestClientObjectMapper` and
+`@RestServerObjectMapper` annotations on your `ObjectMapperCustomizer`.  `ObjectMapperCustomizer` instances without any
+qualifier will customize all three `ObjectMapper`s. `ObjectMapperCustomizer` annotated with `@RestClientObjectMapper` will customize
+only the client instance.  `ObjectMapperCustomizer` annotated with `@RestServerObjectMapper` will customize only the server instance.
```

Comment:
Hmmmm. With the way it's documented, I would have expected the client and server instances to be created automatically if there are customizers and AFAICS from the code, it's not the case?

I.e. I would have expected the third part of the doc where you have to produce the `ObjectMapper` yourselves to not be necessary as it's quite brittle and you could definitely get it wrong doing things by yourself.
So basically, I would have produced synthetic beans for these object mappers if the annotations were found on customizers as IIRC, it's not possible to return null if we want to produce a `@Singleton`.

@geoand does it make sense or am I missing something?

Reply:
Prior comment from some reviewer asked to remove creation of client and server instances.  My quarkus mojo may not be up to the task of creating the beans as part of build.

Reply:
So we don't want to create the ``ObjectMapper``s if they are not needed. My question was: could we create them only if they are required. I think we could do that with synthetic beans but I want to check with @geoand that this makes sense.

Reply:
Hum. And I'm not really sure this works because IIRC the Client writer is only initialized if the server part is not around: https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-client-reactive-jackson/deployment/src/main/java/io/quarkus/rest/client/reactive/jackson/deployment/RestClientReactiveJacksonProcessor.java#L42 .

Also the `JacksonBasicMessageBodyReader` that you haven't changed and should probably be changed is shared between the two.

So my guess is that this needs more thoughts. But I might be missing something as I'm not that familiar with RESTEasy Reactive.

Reply:
@gsmet so basically you are asking if we can create the "RestServerObjectMapper" if there are customizers for it?

Reply:
@geoand so that was my initial question yes. As otherwise it's very cumbersome (especially, since you have to apply the customizers yourself).

But... my other question is: I'm pretty sure the current proposal won't work because in some cases we share the writer between client and server and if I'm not mistaken the reader is always shared.
I'm wondering if we could use `@ConstrainedTo(RuntimeType.SERVER)` / `@ConstrainedTo(RuntimeType.CLIENT)` but I'm not sure it's implemented in RESTEasy Reactive.
If we can, we would make the reader/writer for client/server totally distinct and this could work.

Reply:
> As otherwise it's very cumbersome (especially, since you have to apply the customizers yourself).

I think it's acceptable price to pay, at least initially until we see if people actually use this feature.

> I'm pretty sure the current proposal won't work because in some cases we share the writer between client and server and if I'm not mistaken the reader is always shared.

What reader and writter are you talking about?

Reply:
Good point, I had totally forgot about that! Indeed we would need to always have both and limit them to client and server